### PR TITLE
Feature: add SQL-based `ContractNegotiationStore`

### LINKED ISSUE #864: SQL: implementation of the `ContractNegotiationStore`
An SQL-based implementation of the `TransferProcessStore` is needed.

### Requirements
1. Exclusive Locking: the `nextForState` method must take into account the exclusive locking mechanism required by the asynchronous nature of the `ContractNegotiationManager`s:
    - `nextForState()` locks the TP
    - `update()` unlocks the TP
    - `find()` ignores the lock
    - there should be a timeout after which the lock is automatically released.
1. `save()`: implement upsert semantics
2. Querying: the query implementation must be flexible enough to accomodate  a `QuerySpec`.
3. Table Schema: `ContractAgreement` should be in a separate table (use FK) as `findContractAgreement()` will have to traverse that FK. 
4. Indexing: As a starting point there thould be indexes on `ContractNegotiation.id`, `ContractAgreement.id` and `ContractNegotiation.correlationId`.

_Please provide a written proposal for requirements 1, 3, 4 and 5 before beginning the implementation._

- [x] Design Proposal: #938 
- [ ] Implementation:

### PR DESCRIPTION
**Pull request by paullatzelsperger**
## What this PR changes/adds

Adds an SQL-based implementation of the `ContractNegotiationStore`

## Why it does that

SQL is required as data backend for Milestone 3.

## Further notes

- added a SQL BOM which pulls in all current SQL implementations.
- implemented some basic tests for the `SqlLeaseContext`
- added a util class `LeaseUtil` for usage in tests, which was introduced also in the `SqlTransferProcessStoreTest`

## Linked Issue(s)

Closes #864 

## Checklist

- [x] added appropriate tests?
- [x] performed checkstyle check locally?
- [x] added/updated copyright headers?
- [x] documented public classes/methods?
- [x] added/updated relevant documentation?
- [x] added relevant details to the changelog? (_skip with label `no-changelog`_)
- [x] formatted title correctly? (_take a look at the [CONTRIBUTING](https://github.com/eclipse-dataspaceconnector/DataSpaceConnector/blob/main/CONTRIBUTING.md#submit-a-pull-request) and [styleguide](https://github.com/eclipse-dataspaceconnector/DataSpaceConnector/blob/main/styleguide.md) for details_)

### CODE REVIEW

File: extensions/sql/build.gradle.kts [hunk 1]
```diff
@@ -0,0 +1,35 @@
+/*
+ *  Copyright (c) 2022 Daimler TSS GmbH
```
Comment (zeier) [inline extensions/sql/build.gradle.kts:2]: Probably a copy/paste error on the headers

Comment (paullatzelsperger) [reply extensions/sql/build.gradle.kts]: yup - fixed

File: extensions/sql/contract-negotiation-store/src/main/java/org/eclipse/dataspaceconnector/sql/contractnegotiation/store/SqlContractNegotiationStore.java [hunk 2]
```diff
@@ -0,0 +1,311 @@
...
+                }))
+                .errorDetail(resultSet.getString(statements.getErrorDetailColumn()))
+                .traceContext(fromJson(resultSet.getString(statements.getTraceContextColumn()), new TypeReference<>() {
+                }))
+                .build();
+    }
+
+    private String toJson(Object object) {
+        try {
+            return objectMapper.writeValueAsString(object);
```
Comment (ndr-brt) [inline extensions/sql/contract-negotiation-store/src/main/java/org/eclipse/dataspaceconnector/sql/contractnegotiation/store/SqlContractNegotiationStore.java:290]: `TypeManager` already offer the `writeValueAsString` that handles checked exception, could be preferred to `ObjectMapper`

File: extensions/sql/transfer-process-store/src/test/java/org/eclipse/dataspaceconnector/sql/transferprocess/store/SqlTransferProcessStoreTest.java [hunk 3]
```diff
@@ -77,17 +75,19 @@ void setUp() throws SQLException {
...
         dataSourceRegistry.register(DATASOURCE_NAME, poolDataSource);
         txManager.registerResource(new DataSourceResource(poolDataSource));
-        store = new SqlTransferProcessStore(dataSourceRegistry, DATASOURCE_NAME, transactionContext, new ObjectMapper(), new PostgresStatements(), CONNECTOR_NAME);
+        var statements = new PostgresStatements();
+        store = new SqlTransferProcessStore(dataSourceRegistry, DATASOURCE_NAME, txManager, new ObjectMapper(), statements, CONNECTOR_NAME);
 
         try (var inputStream = getClass().getClassLoader().getResourceAsStream("schema.sql")) {
             var schema = new String(Objects.requireNonNull(inputStream).readAllBytes(), StandardCharsets.UTF_8);
-            transactionContext.execute(() -> executeQuery(connection, schema));
+            ((TransactionContext) txManager).execute(() -> executeQuery(connection, schema));
```
Comment (ndr-brt) [inline extensions/sql/transfer-process-store/src/test/java/org/eclipse/dataspaceconnector/sql/transferprocess/store/SqlTransferProcessStoreTest.java:83]: unneeded cast

**Review by ndr-brt — APPROVED** [review]

**Comment by ndr-brt** [comment]
@jimmarino can we merge this? As it is mandatory for some Data Management Api issues

**Pull request merged**